### ISSUE TITLE
`libnppc.so.12.0.0.30`: ELF load command address/offset not properly aligned

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/libnpp-feedstock#2]
When creating an environment like so...

```bash
conda create -n tst python libnpp -y
conda activate tst
```

...and running the following test...

```bash
cd "${CONDA_PREFIX}"
python -c "import ctypes; ctypes.CDLL('./targets/x86_64-linux/lib/libnppc.so.12.0.0.30')"
```

...am seeing the following issue:

```python
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/datasets/jkirkham/miniconda/envs/tst/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: ./targets/x86_64-linux/lib/libnppc.so.12.0.0.30: ELF load command address/offset not properly aligned
```

<hr>

<details>

<summary>
Result from <code>conda list</code>:
</summary>

```bash
# packages in environment at /datasets/jkirkham/miniconda/envs/tst:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2023.5.7             hbcca054_0    conda-forge
cuda-version              12.0                 hffde075_2    conda-forge
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libexpat                  2.5.0                hcb278e6_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 13.1.0               he5830b7_0    conda-forge
libgomp                   13.1.0               he5830b7_0    conda-forge
libnpp                    12.0.0.30            h59595ed_0    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libsqlite                 3.42.0               h2797004_0    conda-forge
libstdcxx-ng              13.1.0               hfd8a6a1_0    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libzlib                   1.2.13               hd590300_5    conda-forge
ncurses                   6.4                  hcb278e6_0    conda-forge
openssl                   3.1.1                hd590300_1    conda-forge
pip                       23.1.2             pyhd8ed1ab_0    conda-forge
python                    3.11.4          hab00c5b_0_cpython    conda-forge
readline                  8.2                  h8228510_1    conda-forge
setuptools                68.0.0             pyhd8ed1ab_0    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
tzdata                    2023c                h71feb2d_0    conda-forge
wheel                     0.40.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
```

</details>

**Comment by jakirkham** [comment]
Tried testing with the `redist` and am seeing the same issue

@adibbley could you please look into this?

**Comment by pbeaucage** [comment]
I dug into this a bit.  This package just redistributes the NVidia libnpp from developer.nvidia.com, but it appears something wonky is happening - specifically, `conda-build` uses `patchelf` to attempt to make the library redistributable, and in the process seems to corrupt it.

If I manually download the library from NVidia, the test above passes (as expected).  That `so` has a sha256 of:
`978c12f2045c5d7d168790507af5936caddb07858310a01472c1d6aa0f6b56ab  libnpp-linux-x86_64-12.0.0.30-archive/lib/libnppc.so.12.0.0.30`

while the one from c-f is:
`df3f04160f6cf76e6c7b2fefe9106ecb5d29ecabb3834d0ccfb099d064028ab2  libnppc.so.12.0.0.30`

Unfortunately, I'm not super familiar with how to reproduce actions within conda-build so can't debug too much further, but I'll note that in a [recent build run](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=715743&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d) has a number of errors related to patching (line 227):
```WARNING :: get_rpaths_raw()=[] and patchelf=['$ORIGIN'] disagree for [...]targets/x86_64-linux/lib/libnppc.so.12.0.0.30 ```

**Comment by jakirkham** [comment]
Thanks @pbeaucage! 🙏

Sorry for missing your comment (think that landed the day after I left for vacation)

So there are a couple things:

1. `patchelf` version `0.18.0` has a bug ( https://github.com/NixOS/patchelf/issues/492 )
2. Even with `patchelf` version `0.17.x`, using the standard conda-build is problematic ( https://github.com/conda-forge/cuda-feedstock/issues/10 )

Regarding 1, we've marked the `patchelf` version `0.18.0` packages broken ( https://github.com/conda-forge/admin-requests/pull/746 ) and discussed the issue with upstream. It isn't resolved upstream yet, but we are working around it for the time being in conda-forge.

Regarding 2, as discussed in the linked issue, we observed a more general RPATH issue in conda-forge packages and resolved it by disabling conda-build's builtin RPATH changes and adding our own. Think this pertains to the warning at the end of your comment. We have addressed this problem across the CTK, this specific package was fixed with PR ( https://github.com/conda-forge/libnpp-feedstock/pull/4 ).

Hope that provides some more context (and addresses the issue). Happy to follow up on anything else that comes up
